osquery/registry/registry.cpp

```diff
@@ -60,14 +60,21 @@ Status RegistryHelperCore::setActive(const std::string& item_name) {
     }
   }
 
+  Status status(0, "OK");
   active_ = item_name;
   // The active plugin is setup when initialized.
   for (const auto& item : osquery::split(item_name, ",")) {
     if (exists(item, true)) {
-      Registry::get(name_, item)->setUp();
+      status = Registry::get(name_, item)->setUp();
+    } else if (exists(item, false) && !Registry::external()) {
+      // If the active plugin is within an extension we must wait.
+      // An extension will first broadcast the registry, then receive the list
+      // of active plugins, active them if they are extension-local, and finally
+      // start their extension socket.
+      status = pingExtension(getExtensionSocket(external_.at(item_name)));
```

Comment:
This seems to be the source of a few races. There was an incorrect assumption that if the plugin appeared in the core's registry, then it was acceptable to proceed assuming the extension's registry was setup. The correct flow requires a check for a valid extension socket. Extensions will only start their receive sockets when their registry is setup. The core however, receives an extension's broadcast before assigning it a UUID.